src/lsp.rs

```diff
@@ -81,6 +81,99 @@ pub fn get_word_from_pos_params<'a>(
     &line_contents[word_start..word_end]
 }
 
+/// Returns a string slice to the filename in doc specified by the position params
+/// if it is one
+pub fn get_filename_from_pos_params<'a>(
+    doc: &'a FullTextDocument,
+    pos_params: &TextDocumentPositionParams,
+) -> &'a str {
+    let is_ident_char = |c: char| c.is_alphanumeric() || c == '_';
+
+    let line_contents = doc.get_content(Some(Range {
+        start: Position {
+            line: pos_params.position.line,
+            character: 0,
+        },
+        end: Position {
+            line: pos_params.position.line,
+            character: u32::MAX,
+        },
+    }));
+
+    let (word_start, word_end) =
+        find_word_at_pos(line_contents, pos_params.position.character as usize);
+    // try walking forward from the filename to the extension
+    if let Some('.') = line_contents.chars().nth(word_end) {
+        if let Some(idx) = line_contents
+            .chars()
+            .enumerate()
+            .skip(word_end + 1)
+            .take_while(|(_i, c)| is_ident_char(*c))
+            .map(|(i, _c)| i)
+            .last()
+        {
+            return &line_contents[word_start..=idx];
+        }
+    }
+    // try walking backward from the extension to the filename
+    let search_start = if word_start > 0 {
+        word_start - 1
+    } else {
+        word_start
+    };
+    if let Some('.') = line_contents.chars().nth(search_start) {
+        if let Some(idx) = line_contents
+            .chars()
+            .rev()
+            .enumerate()
+            .skip_while(|(i, _c)| (line_contents.len() - 1 - *i) >= word_start)
+            .skip(1)
+            .take_while(|(_i, c)| is_ident_char(*c))
+            .map(|(i, _c)| line_contents.len() - 1 - i)
+            .last()
+        {
+            return &line_contents[idx..word_end];
+        }
+    }
+
+    // could be a file without an extension....
+    &line_contents[word_start..word_end]
+}
+
+/// return a vector of include directories
+pub fn get_include_dirs() -> Vec<PathBuf> {
+    let mut include_dirs = HashSet::new();
+    let cmds = &["cpp", "cpp", "clang", "clang"];
```

Comment:
Just to verify, does this work in cases where the `cpp` or `clang` commands don't exist on the system?